src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs

```diff
@@ -152,5 +168,195 @@ private void PrepareZIndex()
+            var line =  GetLineTextAtCaretsIndex(text,caretIndex);
+            var caretAtLine = GetCaretIndexRelativeToLine(text, caretIndex);
+
+            // If TAB was pressed just after a bullet insert it before the bullet
+            // so that you can easily indent
+            var caretIsRightAfterBullet = IsCaretRightAfterBullet(line, caretAtLine);
+            if (caretIsRightAfterBullet)
+                line = line.Insert(caretAtLine - 2, "\t");
+            else
+            {
+                line = line.Insert(caretAtLine, "\t");
+                caretAtLine++;
+            }
+
+            if (IsCaretBeforeABullet(line, caretAtLine) || caretIsRightAfterBullet)
+                line = UpdateBulletAccordingToIndentation(line);
+            
+            return ReplaceLineOfText(text, lineNumber, line);
+        }
+
+        /// <summary>
+        /// Handles text when user types Key.Enter (ENTER)
+        /// </summary>
+        /// <param name="text">  text before typing ENTER </param>
+        /// <param name="caretIndex"> caret index where user typed ENTER </param>
+        /// <returns></returns>
+        private string BulletEnterHandler(string text, int caretIndex)
+        {
+            var lineNumber = GetLineNumberAtCaretsIndex(text, caretIndex);
+            var line = GetLineTextAtCaretsIndex(text, caretIndex);
+            var caretAtLine = GetCaretIndexRelativeToLine(text, caretIndex);
+
+            // Check if line has any bullets, and copy it to next line
+            var bulletsInLine = BULLETS_CHARS.Where(b => line.Contains(b));
+            if (bulletsInLine.Count() == 0)
+                return text = text.Insert(caretIndex, "\n");
+
+            text = text.Insert(caretIndex, "\n");
+            caretIndex++;
+
+            var tabsBeforeBullet = line.Split(BULLETS_CHARS)[0].Count(c => c=='\t');
+            for (int i = 0; i < tabsBeforeBullet; i++)
+            {
+                text = text.Insert(caretIndex, "\t");
+                caretIndex++;
+            }
+            return text.Insert(caretIndex, bulletsInLine.First()+" ");
+        }
+
+        private bool StringIsEmptyOrTab(string text)
+        {
+            return !text.Any(c => !(c == ' ' || c == '\t'));
+        }
+
+        private bool IsCaretRightAfterBullet(string text, int caretIndex)
+        {
+            if (caretIndex == 0 )
+                return false;
+            return BULLETS_CHARS.Contains(text[caretIndex - 2]);
+        }
+
+        private bool IsCaretBeforeABullet(string text, int caretIndex)
+        {
+           var charactersBeforeBullet = text.Split(BULLETS_CHARS)[0];
+           return caretIndex <= charactersBeforeBullet.Length;
+        }
+
+        private int GetCaretIndexRelativeToLine(string text, int caretIndex)
+        {
+            // Get the text before the caret, split it into lines
+            // The caret index relative to line will be the last line's length
```

Comment:
Maybe just me but not sure i understand this

Reply:
Let me know if this is better